container/shim/src/index.js

```diff
@@ -42,161 +47,231 @@ const PROXY_RESPONSE_HEADERS = [
-  debug.extend('important')('===== IMPORTANT =====')
-
-  for (let i = 0; i < cpus().length; i++) {
-    cluster.fork()
+debug('Saturn L1 Node')
+debug.extend('id')(nodeId)
+debug.extend('version')(NODE_VERSION)
+debug.extend('important')('===== IMPORTANT =====')
+debug.extend('important')(`Earnings will be sent to Filecoin wallet address: ${FIL_WALLET_ADDRESS}`)
+debug.extend('important')(NODE_OPERATOR_EMAIL ? `Payment notifications and important update will be sent to: ${NODE_OPERATOR_EMAIL}` : 'NO OPERATOR EMAIL SET, WE HIGHLY RECOMMEND SETTING ONE')
+debug.extend('important')('===== IMPORTANT =====')
+
+process.on('SIGQUIT', shutdown)
+process.on('SIGINT', shutdown)
+
+setTimeout(async function () {
+  if (argv.register !== false) {
+    await register(true).catch(err => {
+      debug(`Failed to register ${err.name} ${err.message}`)
+      process.exit(1)
+    })
   }
 
-  cluster.on('exit', () => {
-    if (Object.keys(cluster.workers).length === 0) {
-      debug('All servers closed')
-      shutdownCluster()
-    }
-  })
+  // Start log ingestor
+  await initLogIngestor()
+}, 500)
 
-  process.on('SIGQUIT', shutdownCluster)
-  process.on('SIGINT', shutdownCluster)
+const ipfsAgent = new https.Agent({
+  keepAlive: true,
+  maxSockets: Math.floor(256 / cpus().length)
+})
 
-  setTimeout(async function () {
-    if (argv.register !== false) {
-      await register(true).catch(err => {
-        debug(`Failed to register ${err.name} ${err.message}`)
-        process.exit(1)
-      })
-    }
+const app = express()
 
-    // Start log ingestor
-    await initLogIngestor()
-  }, 500)
-} else {
-  const ipfsAgent = new https.Agent({
-    keepAlive: true,
-    maxSockets: Math.floor(256 / cpus().length)
-  })
+const testCAR = await fsPromises.readFile('./public/QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF.car')
 
-  const app = express()
+const connectedL2Nodes = new Map()
+const openCARRequests = new Map()
 
-  const testCAR = await fsPromises.readFile('./public/QmQ2r6iMNpky5f1m4cnm3Yqw8VSvjuKpTcK1X7dBR1LkJF.car')
+function removeConnectedL2Node (id) {
+  const { res, cleanedUp } = connectedL2Nodes.get(id)
+  cleanedUp.value = true
+  res.end()
+  connectedL2Nodes.delete(id)
+}
 
-  app.disable('x-powered-by')
-  app.set('trust proxy', true)
+app.disable('x-powered-by')
+app.set('trust proxy', true)
 
-  app.get('/favicon.ico', (req, res) => {
-    res.sendStatus(404)
-  })
+app.get('/favicon.ico', (req, res) => {
+  res.sendStatus(404)
+})
+
+// Whenever nginx doesn't have a CAR file in cache, this is called
+app.get('/ipns/:cid', handleCID)
+app.get('/ipns/:cid/:path*', handleCID)
+app.get('/ipfs/:cid', handleCID)
+app.get('/ipfs/:cid/:path*', handleCID)
 
-  // Whenever nginx doesn't have a CAR file in cache, this is called
-  app.get('/ipns/:cid', handleCID)
-  app.get('/ipns/:cid/:path*', handleCID)
-  app.get('/ipfs/:cid', handleCID)
-  app.get('/ipfs/:cid/:path*', handleCID)
+async function handleCID (req, res) {
```

Comment:
Express does not support async request handlers (yet another reason why to use Fastify).

If this async function returns a promise that's eventually rejected, unhandled promise rejection is triggered at the process level. I don't remember whether this prints a warning or crashes the process. I think neither is desirable.

A possible solution: convert the async hander into callback-based.

```js
function wrap(fn) {
  return function(req, res, next) {
    fn(req, res)
      .catch(next)
      .catch(
        // crash if callback throws
        err => process.nextTick(() => { throw err; })
      });
  };
}

const handleCID = wrap(async function handleCIDimpl(req, res) {
  // your code
}
```

Reply:
this package is often recommended to handle this issue https://www.npmjs.com/package/express-async-errors

Reply:
Quoting from https://www.fastify.io/benchmarks/
- Fastify - 53913 req/sec
- Express - 11739 req/sec

On the second thought, I think the Node.js HTTP framework is not a bottleneck in our current L1 implementation since most CPU is consumed by Nginx anyways.

I am fine to stay with Express for now. To me, it's important to have robust & reliable error handling. This can be achieved with Express, too: using one of the plugins like the one mentioned above, wrapping our handler to convert them to callback style, or we can even rewrite the handlers into callback style.

Reply:
I agree, and captured this in https://github.com/filecoin-saturn/L1-node/issues/33, which we should also resolve before launch.